File: java/src/main/java/fr/sdecout/handson/rest/shared/ValidIsbn.java
package fr.sdecout.handson.rest.shared;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Documented
@Constraint(validatedBy = ValidIsbnValidator.class)
@Target({FIELD, PARAMETER})
@Retention(RUNTIME)
public @interface ValidIsbn {
    String message() default "Value is not a valid ISBN";
    Class<?>[] groups() default {};
    Class<? extends Payload>[] payload() default {};
}

File: kotlin/src/main/kotlin/fr/sdecout/handson/rest/shared/ValidIsbn.kt
package fr.sdecout.handson.rest.shared

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.FIELD
import kotlin.annotation.AnnotationTarget.VALUE_PARAMETER
import kotlin.reflect.KClass

@MustBeDocumented
@Constraint(validatedBy = [ValidIsbnValidator::class])
@Target(FIELD, VALUE_PARAMETER)
@Retention(RUNTIME)
annotation class ValidIsbn(
    val message: String = "Value is not a valid ISBN",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
